### PR TITLE
Add test for dying with dignity

### PR DESCRIPTION
**Pull request by jasontedor**
I have long wanted an actual test that dying with dignity works. It is tricky because if dying with dignity works, it means the test JVM dies which is usually an abnormal condition. And anyway, how does one force a fatal error to be thrown. I was motivated to investigate this again by the fact that I missed a backport to one branch leading to an issue where Elasticsearch would not successfully die with dignity. And now we have a solution: we install a plugin that throws an out of memory error when it receives a request. We hack the standalone test infrastructure to prevent this from failing the test. To do this, we bypass the security manager and remove the PID file for the node; this tricks the test infrastructure into thinking that it does not need to stop the node. We also bypass seccomp so that we can fork jps to make sure that Elasticsearch really died. And to be extra paranoid, we parse the logs of the dead Elasticsearch process to make sure it died with dignity. Never forget.

Relates #19272

### CODE REVIEW

**Review by hub-cap — COMMENTED** [review]
LGTM but I would like another human to review

File: qa/die-with-dignity/build.gradle [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+apply plugin: 'elasticsearch.esplugin'
+
+esplugin {
+    description 'Out of memory plugin'
+    classname 'org.elasticsearch.DieWithDignityPlugin'
+}
+
+integTestCluster {
```
Comment (martijnvg) [inline qa/die-with-dignity/build.gradle]: is this empty block needed?

**Review by martijnvg — APPROVED** [review]
LGTM, nice to see how it is still possible to test this kind of stuff!

Too bad the PR build failed, because a JRE is installed instead of a JDK :(

File: qa/die-with-dignity/src/test/java/org/elasticsearch/qa/die_with_dignity/DieWithDignityIT.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    public void testDieWithDignity() throws Exception {
+        // deleting the PID file prevents stopping the cluster from failing since it occurs if and only if the PID file exists
+        final Path pidFile = PathUtils.get(System.getProperty("pidfile"));
+        final List<String> pidFileLines = Files.readAllLines(pidFile);
+        assertThat(pidFileLines, hasSize(1));
+        final int pid = Integer.parseInt(pidFileLines.get(0));
+        Files.delete(pidFile);
+        final CountDownLatch latch = new CountDownLatch(1);
+        client().performRequestAsync("GET", "/_die_with_dignity", new ResponseListener() {
```
Comment (nik9000) [inline qa/die-with-dignity/src/test/java/org/elasticsearch/qa/die_with_dignity/DieWithDignityIT.java]: Why not `performRequest`?

**Review by nik9000 — APPROVED** [review]

File: qa/die-with-dignity/src/test/java/org/elasticsearch/qa/die_with_dignity/DieWithDignityIT.java (hunk 2 above)
Comment (jasontedor) [reply qa/die-with-dignity/src/test/java/org/elasticsearch/qa/die_with_dignity/DieWithDignityIT.java]: It's a long story how it ended up this way that is not interesting, but now it is no longer needed. It's leftover from a previous iteration. I pushed 14c720edabbf0e8a87b82d595f7592366d7f0e3a.

File: qa/die-with-dignity/build.gradle (hunk 1 above)
Comment (jasontedor) [reply qa/die-with-dignity/build.gradle]: I pushed 881a8d21a9e251046e651a6ea3561db5f1bf996a.

**Review by nik9000 — APPROVED** [review]

**Review by rjernst — COMMENTED** [review]
The test looks fine, I just have one comment about the name. I know we have used this phrase internally to mean "let the node die when an error occurs", but thus far we have only used it in PR titles, AFAIK. I think for actual code, we should use a more plain name that will not be lost in translation, or be only known to those that have been around working on ES in the 2.x-6.x range. I think something like "qa/error-exit" or "qa/exit-on-error" would be much more clear to anyone browsing through the qa tests who did not know this "insider" phrase.

**Comment by jasontedor** [comment]
I disagree, because we refer to this as "die with dignity" everywhere. It has appeared in [release notes](https://www.elastic.co/blog/elasticsearch-5-0-2-released) too. And if ever we were to write a blog post about why we do this, it would definitely refer to this as "dying with dignity". It's a fun name.

**Pull request merged**